code/datums/components/riding.dm

```diff
@@ -0,0 +1,309 @@
+	AM.pixel_x = px
+	AM.pixel_y = py
+
+/datum/component/riding/proc/set_vehicle_dir_offsets(dir, x, y)
+	directional_vehicle_offsets["[dir]"] = list(x, y)
+
+//Override this to set your vehicle's various pixel offsets
+/datum/component/riding/proc/get_offsets(pass_index) // list(dir = x, y, layer)
+	. = list(TEXT_NORTH = list(0, 0), TEXT_SOUTH = list(0, 0), TEXT_EAST = list(0, 0), TEXT_WEST = list(0, 0))
+	if(riding_offsets["[pass_index]"])
+		. = riding_offsets["[pass_index]"]
+	else if(riding_offsets["[RIDING_OFFSET_ALL]"])
+		. = riding_offsets["[RIDING_OFFSET_ALL]"]
+
+/datum/component/riding/proc/set_riding_offsets(index, list/offsets)
+	if(!islist(offsets))
+		return FALSE
+	riding_offsets["[index]"] = offsets
+
+//Override this to set the passengers/riders dir based on which passenger they are.
+//ie: rider facing the vehicle's dir, but passenger 2 facing backwards, etc.
+/datum/component/riding/proc/get_rider_dir(pass_index)
+	var/atom/movable/AM = parent
+	return AM.dir
+
+//KEYS
+/datum/component/riding/proc/keycheck(mob/user)
+	return !keytype || istype(user.get_held_item(), keytype)
+
+//BUCKLE HOOKS
+/datum/component/riding/proc/restore_position(mob/living/buckled_mob)
+	if(buckled_mob)
+		buckled_mob.pixel_x = 0
+		buckled_mob.pixel_y = 0
+		if(buckled_mob.client)
+			buckled_mob.client.change_view(CONFIG_GET(string/default_view))
+
+//MOVEMENT
+/datum/component/riding/proc/turf_check(turf/next, turf/current)
+	if(allowed_turf_typecache && !allowed_turf_typecache[next.type])
+		return (allow_one_away_from_valid_turf && allowed_turf_typecache[current.type])
+	else if(forbid_turf_typecache && forbid_turf_typecache[next.type])
+		return (allow_one_away_from_valid_turf && !forbid_turf_typecache[current.type])
+	return TRUE
+
+/datum/component/riding/proc/handle_ride(mob/user, direction)
+	var/atom/movable/AM = parent
+	if(user.incapacitated())
+		Unbuckle(user)
+		return
+
+	if(world.time < last_vehicle_move + ((last_move_diagonal ? 2 : 1) * vehicle_move_delay))
+		return
+	last_vehicle_move = world.time
+
+	if(keycheck(user))
+		var/turf/next = get_step(AM, direction)
+		var/turf/current = get_turf(AM)
+		if(!istype(next) || !istype(current))
+			return	//not happening.
+		if(!turf_check(next, current))
+			to_chat(user, "<span class='warning'>Your \the [AM] can not go onto [next]!</span>")
+			return
+		if(!isturf(AM.loc))
+			return
+		step(AM, direction)
+
+		if((direction & (direction - 1)) && (AM.loc == next))		//moved diagonally
+			last_move_diagonal = TRUE
+		else
+			last_move_diagonal = FALSE
+
+		handle_vehicle_layer()
+		handle_vehicle_offsets()
+	else
+		to_chat(user, "<span class='warning'>You'll need the keys in one of your hands to [drive_verb] [AM].</span>")
+
+/datum/component/riding/proc/Unbuckle(atom/movable/buckled_thing)
+	addtimer(CALLBACK(parent, /atom/movable/.proc/unbuckle_mob, buckled_thing), 0, TIMER_UNIQUE)
+
+///////Yes, I said humans. No, this won't end well...//////////
+/datum/component/riding/human
```

Comment:
why does this need a subtype

Reply:
It's a special case, directly taken from /tg/
It has several redefinitions from the base. Not making it a subtype would require quite the refactor.